[mlir][IR] Add `InsertPoint::after(ValueRange)`

matthias-springer opened the pull request:
This commit adds a helper function `InsertPoint::after`. The function computes an insertion point that post-dominates the definition of all given values.

This functionality will be used in the merged 1:1 / 1:N dialect conversion to determine where to insert a `builtin.unrealized_conversion_cast` op. To ensure that the same cast can be "reused" for multiple consumer ops, the cast should be inserted as early as possible. (Alternatively, multiple cast ops could be inserted, right before each consumer op.)

TODO: Add test case.

Depends on #115587.

matthias-springer commented:
I could put this helper function into `DialectConversion.cpp`, but I thought it could be useful helper in general. (I'm also trying to make the 1:N dialect conversion PR smaller before sending it for review.) What's unclear is how to test this. Maybe a unit test? What do you think?

River707 commented on mlir/lib/IR/Builders.cpp:
```diff
@@ -641,3 +641,76 @@ void OpBuilder::cloneRegionBefore(Region &region, Region &parent,
+}
+
+OpBuilder::InsertPoint
+OpBuilder::InsertPoint::findClosest(ArrayRef<Value> values) {
```
This is a fairly complex bit of logic here that isn't tested, could this be separated out from the `after` logic?

Also, I would have expected the internals to use DominanceInfo/DominanceFrontier/etc. These types of traversals are going to be quite slow for anything other than the simple case otherwise.

matthias-springer committed:
[mlir][IR][NFC] `DominanceInfo`: Minor code cleanups

matthias-springer committed:
[mlir][IR] `DominanceInfo`: Deduplicate `properlyDominates` implementation

The implementations of `DominanceInfo::properlyDominates` and `PostDominanceInfo::properlyPostDominates` are almost identical: only one line of code is different. Define the function in `DominanceInfoBase` to avoid the code duplication.

Note: This commit is not marked as NFC because `PostDominanceInfo::properlyPostDominates` now also has an `enclosingOpOk` argument.

matthias-springer commented:
I'm trying to implement this in a different way: based on dominance, as you suggested. By adding a new `DominanceInfo::dominates(Block*, Block::iterator, Block*, Block::iterator)` function. Still experimenting with the design... Marking this PR as "draft" until then.

joker-eph commented on mlir/include/mlir/IR/Builders.h:
```diff
@@ -334,6 +334,18 @@ class OpBuilder : public Builder {
   /// This class represents a saved insertion point.
   class InsertPoint {
   public:
+    /// Finds the closest insertion point where all given values are defined
```
It's not clear to me that this is a well defined problem: what does "closest means" here?

matthias-springer replied:
Possible candidates are the points (`Block::iterator`) right after the definition of each value. I will take a look at all of these and find the one that post-dominates all the others. I think there should be only one such point. Will update this PR accordingly soon...

joker-eph commented on mlir/lib/IR/Builders.cpp:
```diff
@@ -641,3 +641,76 @@ void OpBuilder::cloneRegionBefore(Region &region, Region &parent,
+    }
+
+    // `pt` and `result` are in different subtrees: neither is an ancestor of
+    // the other. In that case, there is no suitable insertion point.
```
Wouldn't the right insertion point be the post-dominator block?

matthias-springer replied:
Yes. Post-dominator `Block::iterator` to be precise. (Adding support for that in #115587.)

matthias-springer committed:
[mlir][IR][NFC] `DominanceInfo`: Share same impl for block/op dominance

The `properlyDominates` implementations for blocks and ops are very similar. This commit replaces them with a single implementation that operates on block iterators. That implementation can be used to implement both `properlyDominates` variants.

Note: A subsequent commit will add a new public `properlyDominates` overload that accepts block iterators. That functionality can then be used to find a valid insertion point at which a range of values is defined (by utilizing post dominance).

Depends on #115433.

matthias-springer committed:
[mlir][IR] Add `OpBuilder::setInsertionPointAfterValues`

matthias-springer commented on mlir/include/mlir/IR/Builders.h:
```diff
@@ -341,6 +342,19 @@ class OpBuilder : public Builder {
+    /// There may be multiple suitable insertion points. This function chooses
+    /// an insertion right after one of the given values.
+    ///
+    /// Note: Some of the given values may already have gone out of scope at the
```
Note: if there exists an insertion point that post-dominates all definitions and where all values are in scope, this function is guaranteed to find it.

River707 commented:
> > Why don't the various transforms just use DominanceInfo? Answering these queries is what it's intended for.
> 
> Block reachability and block dominance are two different things. Just because there is a path from `block A -> block B` does not necessarily mean that `block A dominates block B`. I'm not sure how "block reachability" could be implemented with `DominanceInfo`.
> 
> That being said, in the dialect conversion I actually need dominance, so what I implemented here is not useful there...

Yes, my comment was geared towards the fact that several of the usages are already conflating reachability with dominance (for example the original usage in the followup IRBuilder PR). My concern is/was around the general utility of such a function, given that it cannot be used in most situations people would reach for it (i.e. dominance outside of purely structured control flow) so it wasn't clear the utility in adding it to Block. That being said I don't have an issue with this singular function, given that it's a tiny amount of graph traversal code. I would clearly mark in the comment that this function is not a replacement for dominance in any reasonable situation (given that a successor block can easily reach a predecessor in a situation such as a loop).

matthias-springer commented:
> Yes, my comment was geared towards the fact that several of the usages are already conflating reachability with dominance (for example the original usage in the followup IRBuilder PR). My concern is/was around the general utility of such a function, given that it cannot be used in most situations people would reach for it (i.e. dominance outside of purely structured control flow) so it wasn't clear the utility in adding it to Block. That being said I don't have an issue with this singular function, given that it's a tiny amount of graph traversal code. I would clearly mark in the comment that this function is not a replacement for dominance in any reasonable situation (given that a successor block can easily reach a predecessor in a situation such as a loop).

I think this comment is referring to #114928. I updated the comment on that PR.

matthias-springer commented:
I don't have any immediate use case for this PR anymore.
